### ISSUE TITLE
Regex causes stack overflow on a very large string

### ISSUE THREAD

**Issue by watzon** [crystal-lang/crystal#8199]
Using the following code:
```crystal
param = File.read("param.txt")
puts param.match(/^(?:[A-Za-z0-9+\/]{4})*(?:[A-Za-z0-9+\/]{2}==|[A-Za-z0-9+\/]{3}=)?$/)
```
With the file [param.txt](https://github.com/crystal-lang/crystal/files/3628977/param.txt) causes a stack overflow
```
Stack overflow (e.g., infinite or very deep recursion)
[0x476ef6] *CallStack::print_backtrace:Int32 +118
[0x4674f0] __crystal_sigfault_handler +192
[0x7f1f58a89c60] ???
[0x7f1f58bf59e4] ???
[0x7f1f58c0463e] ??? (17446 times)
[0x7f1f58bf8db6] ???
[0x7f1f58c076f6] pcre_exec +2710
[0x4d2258] *Regex#match_at_byte_index<String, Int32, Regex::Options>:(Regex::MatchData | Nil) +504
[0x4d1e90] *Regex#match<String, Int32>:(Regex::MatchData | Nil) +176
[0x4813ed] *String#match<Regex>:(Regex::MatchData | Nil) +77
[0x454a19] __crystal_main +4329
[0x4f2083] *Crystal::main_user_code<Int32, Pointer(Pointer(UInt8))>:Nil +35
[0x4f1fc9] *Crystal::main<Int32, Pointer(Pointer(UInt8))>:Int32 +41
[0x45ff36] main +6
[0x7f1f5884df43] __libc_start_main +243
[0x45386e] _start +46
[0x0] ???
```

I'm sure this can be solved by doing some pre-processing, but it seems like something that could be a potential bug and security vulnerability so I figured I'd drop it here

**Comment by watzon** [comment]
Did some checking. The last place at which the expression works without overflowing on my system is 69784 characters in. Don't know if that's helpful or not, but it's something.

```crystal
# This works
param = param[-69784..-1]
```

**Comment by bararchy** [comment]
@bcardiff, I messaged you about this issue a few month ago, we didn't have a proper way to recreate it then but it seems we do now thanks to @watzon.
This should be taken care of as a Critical level issue, in the most naive situation this means a DOS attack on crystal application which regex data from users, in a less naive situation, this might be weaponized into something more malicious.

**Comment by watzon** [comment]
Exactly. Any easily created stack overflow is a potentially pretty serious security vulnerability. The fix in this case is as easy as shortening the string to less than 69000 characters, but that won't work in all cases.

**Comment by refi64** [comment]
If you're taking any sort of untrusted input, PCRE regexes can generally be
unsafe anyway: https://swtch.com/~rsc/regexp/regexp1.html

**Comment by bararchy** [comment]
@refi64 That's true, but there is a "this regex does so many things it will make CPU go 100%" and there is "using regex will cause a stack overflow" :man_shrugging:

**Comment by watzon** [comment]
Yeah, using up the CPU's resources may be inevitable, but a stack overflow in the standard library is an issue

**Comment by oprypin** [comment]
I think that Google concluded many years ago that PCRE is not possible to defend against user input and implemented RE2.
https://github.com/google/re2/wiki/WhyRE2

**Comment by asterite** [comment]
It seems to be dying in Pcre code, not Crystal code.

You could try using Ruby and see what happens. Ruby uses onigmo, if that works well we can consider using that instead.

Otherwise, yes, RE2 might be the safest alternative.

**Comment by watzon** [comment]
Seems to work just fine in Ruby @asterite. Personally I'd love to see Onigmo used in Crystal over PCRE. I know Ruby compatibility isn't a Crystal goal, but being able to search Ruby questions and adapt the answer to Crystal is definitely one of it's benefits and having them use the same regex engine would help that.

**Comment by wdhwg001** [comment]
For alternatives, I would also suggest Onigmo rather than RE2. It has more features like conditions and recursions. I believe that Onigmo can be the most feature-rich regex engine.

**Comment by asterite** [comment]
Can you provide the contents of "param.txt" so I can check whether oniguruma doesn't choke on it?

**Comment by bararchy** [comment]
@asterite -> https://github.com/crystal-lang/crystal/files/3628977/param.txt

**Comment by asterite** [comment]
It seems to work fine with oniguruma.

But still, maybe re2 is the safest.

**Comment by wdhwg001** [comment]
@asterite RE2's linear-time performance comes with a cost:

All recursive, conditional, look-behind and look-forward syntax are not supported. (https://github.com/google/re2/wiki/Syntax)

IMO, crystal is a language for general purpose instead of only for server-side development. Onigmo has generally the same performance as RE2, and if the developer don't use any recursive / look-behind / look-forward feature, the performance should also be linear. For me, I would love to choose a full-feature regex engine instead of a probably safer but feature-limited one.

**Comment by asterite** [comment]
@wdhwg001 Definitely! 👍 

Do you know what are the differences between oniguruma and onigmo? I only tried oniguruma because I found it easier to install and link.

**Comment by RX14** [comment]
I'm half tempted to evaluate rure - the C API for rust's regex create - for this usecase too. My major concern would be the fact that `rure` is not packaged very often.

**Comment by naqvis** [comment]
Though bit outdated, but believe comparison of regex engines still valid for further investigation.
[https://rust-leipzig.github.io/regex/2017/03/28/comparison-of-regex-engines/](https://rust-leipzig.github.io/regex/2017/03/28/comparison-of-regex-engines/)

I'll vote for PCRE2 as believe implementation and transition will be quite smooth.

**Comment by watzon** [comment]
@naqvis the question would be, does PCRE2 fix the issues with PCRE? Personally I think Oniguruma is the way to go. It may require updating existing codebases, but the changes should be fairly minimal and the performance increase should be pretty significant.

**Comment by asterite** [comment]
In my experiments oniguruma was slower than PCRE, and MatchData needs to become a class instead of a struct because it needs to be finalized (in PCRE you can set the memory allocator, doesn't seem to be the case with oniguruma).

**Comment by wdhwg001** [comment]
@asterite Onigmo is the choice of Ruby 2.0+.

Onigmo and Oniguruma are nearly the same, except the support of callouts and ASCII/Unicode handling:

Oniguruma supports Callouts like `(?{...contents...}[tag]D)` and `(*name[tag]{args...})`, but Onigmo does not. (Perl supports regex callouts)

For ASCII/Unicode handling, Oniguruma supports separate on/off switches for ASCII word, ASCII digits and ASCII spaces while Onigmo has only ASCII/Unicode switch.

Additionally, they all support Absent functions like `(?~|absent|exp)`. Both implementations seem to be experimental, but Oniguruma has slightly better documentation.

See:
https://github.com/kkos/oniguruma/blob/master/doc/RE
https://github.com/k-takata/Onigmo/blob/master/doc/RE

I haven't tested the performance.

**Comment by RX14** [comment]
2017 is a while ago, I suggest we re-benchmark all three, 

is there a library of representitive testcases to benchmark on available anywhere?

**Comment by naqvis** [comment]
Yes @watzon, provided regex and input works fine in PCRE2. Tested using below golang code

```go
import (
	"fmt"
	"io/ioutil"
	"os"

	"github.com/lestrrat/go-pcre2"
)

func main() {
	re := pcre2.MustCompile(`^(?:[A-Za-z0-9+\/]{4})*(?:[A-Za-z0-9+\/]{2}==|[A-Za-z0-9+\/]{3}=)?$`)
	src, err := ioutil.ReadFile("./param.txt")
	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}

	for i, match := range re.FindAllString(string(src), -1) {
		fmt.Println(match, "found at index", i)
	}
}
```
Same test with PCRE library fails with **seg fault**

**Comment by asterite** [comment]
You can also use my oniguruma branch for Crystal: https://github.com/asterite/crystal-1/tree/oniguruma

You can only `match` for now and dont `p` the match, just do `match[0]`, `match[1]`, etc., because not everything is implemented. But using `match` can be enough to test the performance against PCRE and others.

Of course, if oniguruma doesn't crash like PCRE but it's slower and provides the same features we should switch to oniguruma.

**Comment by funny-falcon** [comment]
Interestingly, but original problem could easily solved if not all work is done with regexp:
```crystal
param = File.read("param.txt")
retutn param =~ /^[A-Za-z0-9+\/]+(={2,3})?$/) && param.length % 4 == 0

**Comment by watzon** [comment]
@funny-falcon of course, and that's basically what I ended up doing, but the regex thing is still a problem

**Comment by juliodcs** [comment]
@watzon, I think you may use your original regex by using [possessive quantifiers](https://www.regular-expressions.info/possessive.html).

`^(?:[A-Za-z0-9+\/]{4})*+(?:[A-Za-z0-9+\/]{2}==|[A-Za-z0-9+\/]{3}=)?$`

To my mind, I'd rather have a feature rich regex engine even if that implies I sometimes need to take care of optimizing the regular expressions.

**Comment by watzon** [comment]
@juliodcs this isn't really about optimizations. This is a potential security vulnerability, and one that's pretty easy to fall in to. The other RegEx engines being discussed aren't really any less feature rich, but they also don't have this issue.

**Comment by juliodcs** [comment]
@watzon, There has been some talking about re2, which does not even support look-ahead, for example.

Onigmo on the other hand, does not seem to support [Backtracking Control Verbs](https://www.rexegg.com/backtracking-control-verbs.html), and I think that the use of `m` flag for `DOTALL` modifier is unfortunate. 

That being said, I think It is great that can take care of killer regexes without crashing. Of course, this protection seem to have some sort of performance cost. 

For example, this (bad) regex:

```
string = ('a' * 19) + 'b'
puts string.match(/^((a)*)*$/)
```

is absolutelly not a problem with php (that uses PCRE2 for version >= 7.3) as it takes milliseconds to process. However in ruby it takes almost 9 minutes. And adding more characters to the string makes it much worse.

Also, I'm just wondering wether one can blame a regex engine for having a stack overflow:

If a stack overflow is raised, probably there's something bad on the regex. The programmer responsability is to fix it, as he/she would do with any other piece of code. I don't see much difference in fixing a bad piece of code vs fixing a regex.

You can also do stack buffer overflows in other ways, not just by regex. The solution is not to use a programming language that forbids recursion, but to fix the code.

In the end whay I mean with all of this is that each engine has strengths and weaknesses. And probably more debate is needed for switching to another regex engine. I think there should be a clear reason/benefit (not just that one regex crashes)
